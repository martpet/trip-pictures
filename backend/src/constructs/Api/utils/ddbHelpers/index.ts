export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
