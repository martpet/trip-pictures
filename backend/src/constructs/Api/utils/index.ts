export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
