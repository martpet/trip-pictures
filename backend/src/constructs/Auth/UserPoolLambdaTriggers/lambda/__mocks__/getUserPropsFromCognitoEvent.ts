export const getUserPropsFromCognitoEvent = vi
  .fn()
  .mockName('getUserPropsFromCognitoEvent')
  .mockReturnValue({
    id: 'dummyId',
    dummyUserPropFromCognitoEventKey: 'dummyUserPropFromCognitoEventValue',
  });
