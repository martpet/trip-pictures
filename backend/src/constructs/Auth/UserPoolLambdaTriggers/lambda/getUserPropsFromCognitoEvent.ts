import {
  CognitoIdentity,
  IdentityProvider,
  PostAuthenticationTriggerEvent,
  PostConfirmationTriggerEvent,
  UserPropsFromCognito,
} from 'lambda-layer';

type TriggerEvent = PostConfirmationTriggerEvent | PostAuthenticationTriggerEvent;

export const getUserPropsFromCognitoEvent = (
  event: TriggerEvent
): UserPropsFromCognito => {
  const { userAttributes } = event.request;
  const identity = JSON.parse(userAttributes.identities)[0] as CognitoIdentity;

  return {
    id: userAttributes.sub,
    providerName: identity.providerName as IdentityProvider,
    givenName: userAttributes.given_name,
    familyName: userAttributes.family_name,
    picture: userAttributes.picture as string | undefined,
    email: userAttributes.email,
    dateCreated: identity.dateCreated,
  };
};
