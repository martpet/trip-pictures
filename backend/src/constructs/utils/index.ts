export * from './createEdgeFunction';
export * from './createNodejsFunction';
export * from './createTable';
