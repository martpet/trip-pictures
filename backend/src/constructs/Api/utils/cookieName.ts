import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
