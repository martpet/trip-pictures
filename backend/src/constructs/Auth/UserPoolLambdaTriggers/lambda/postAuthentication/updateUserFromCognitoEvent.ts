import {
  createDynamoUpdateExpression,
  DynamoDBClient,
  DynamoDBDocumentClient,
  filterChangedProps,
  GetCommand,
  PostAuthenticationTriggerEvent,
  UpdateCommand,
  UserPropsFromCognito,
  UsersTableItem,
  usersTableOptions,
} from 'lambda-layer';

import { getUserPropsFromCognitoEvent } from '../getUserPropsFromCognitoEvent';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient, {
  marshallOptions: { removeUndefinedValues: true },
});

export const updateUserFromCognitoEvent = async (
  event: PostAuthenticationTriggerEvent
) => {
  const propsFromEvent = getUserPropsFromCognitoEvent(event);
  const user = await fetchUser(propsFromEvent);

  if (!user) {
    throw new Error(`could not find user with id "${propsFromEvent.id}"`);
  }

  const changedProps = filterChangedProps(propsFromEvent, user);

  if (!changedProps) {
    return undefined;
  }

  const updateCommand = new UpdateCommand({
    ...makeTableParams(user),
    ...createDynamoUpdateExpression(changedProps),
  });

  return ddbDocClient.send(updateCommand);
};

async function fetchUser(props: UserPropsFromCognito) {
  const getCommand = new GetCommand(makeTableParams(props));
  const { Item } = await ddbDocClient.send(getCommand);
  return Item;
}

function makeTableParams(props: Partial<UsersTableItem>) {
  const pkName = usersTableOptions.partitionKey.name;
  return {
    TableName: usersTableOptions.tableName,
    Key: { [pkName]: props[pkName] },
  };
}
