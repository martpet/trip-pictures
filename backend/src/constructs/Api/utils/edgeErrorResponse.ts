import { getReasonPhrase } from 'http-status-codes';

import { LambdaEdgeResponse } from '~/constructs/Api/types';
import { errorResponse } from '~/constructs/Api/utils/errorResponse';

export const edgeErrorResponse = (
  ...errorResponseArgs: Parameters<typeof errorResponse>
): LambdaEdgeResponse => {
  const { statusCode, body } = errorResponse(...errorResponseArgs);

  const response: LambdaEdgeResponse = {
    status: statusCode,
    statusDescription: getReasonPhrase(statusCode),
    headers: { 'content-type': [{ value: 'application/json' }] },
    body,
  };

  return response;
};
