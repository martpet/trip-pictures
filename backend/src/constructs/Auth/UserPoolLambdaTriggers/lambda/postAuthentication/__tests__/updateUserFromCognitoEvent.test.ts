import { DynamoDBDocumentClient, GetCommand, UpdateCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';

import {
  itCalls,
  itRejects,
  itResolves,
  itSendsAwsCommand,
} from '~/constructs/Api/utils';
import { createDynamoUpdateExpression, filterChangedProps } from '~/utils';

import { getUserPropsFromCognitoEvent } from '../../getUserPropsFromCognitoEvent';
import { updateUserFromCognitoEvent } from '../updateUserFromCognitoEvent';
import event from './__fixtures__/postAuthenticationEvent';

vi.mock('../../getUserPropsFromCognitoEvent');
vi.mock('~/../../shared/utils/filterChangedProps');
vi.mock('~/utils/createDynamoUpdateExpression');

const ddbMock = mockClient(DynamoDBDocumentClient);

const args = [event] as Parameters<typeof updateUserFromCognitoEvent>;

beforeEach(() => {
  ddbMock.reset();

  ddbMock.on(GetCommand).resolves({
    Item: {
      id: 'dummyId',
      dummyGetCommandOutputKey: 'dummyGetCommandOutputValue',
    },
  });

  ddbMock.on(UpdateCommand).resolves({
    Attributes: { dummyUpdateCommandAttrKey: 'dummyUpdateCommandAttrValue' },
  });
});

describe('updateUserFromCognitoEvent', () => {
  itCalls(getUserPropsFromCognitoEvent, updateUserFromCognitoEvent, args);
  itCalls(filterChangedProps, updateUserFromCognitoEvent, args);
  itSendsAwsCommand(GetCommand, ddbMock, updateUserFromCognitoEvent, args);

  describe('when output from "GetCommand" does not include "Item"', () => {
    beforeEach(() => {
      ddbMock.on(GetCommand).resolves({});
    });
    itRejects(updateUserFromCognitoEvent, args);
  });

  describe('when user props in event are not changed', () => {
    beforeEach(() => {
      vi.mocked(filterChangedProps).mockImplementationOnce(() => undefined);
    });
    itResolves(updateUserFromCognitoEvent, args);
  });

  describe('when user props in event are changed', () => {
    itCalls(createDynamoUpdateExpression, updateUserFromCognitoEvent, args);
    itSendsAwsCommand(UpdateCommand, ddbMock, updateUserFromCognitoEvent, args);
    itResolves(updateUserFromCognitoEvent, args);
  });
});
