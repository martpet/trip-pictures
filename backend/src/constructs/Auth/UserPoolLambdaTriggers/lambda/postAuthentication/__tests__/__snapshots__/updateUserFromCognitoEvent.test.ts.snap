// Vitest Snapshot v1

exports[`updateUserFromCognitoEvent > calls "filterChangedProps" with correct args 1`] = `
[
  [
    {
      "dummyUserPropFromCognitoEventKey": "dummyUserPropFromCognitoEventValue",
      "id": "dummyId",
    },
    {
      "dummyGetCommandOutputKey": "dummyGetCommandOutputValue",
      "id": "dummyId",
    },
  ],
]
`;

exports[`updateUserFromCognitoEvent > calls "getUserPropsFromCognitoEvent" with correct args 1`] = `
[
  [
    {
      "callerContext": {
        "awsSdkVersion": "dummyPostAuthAwsSdkVersion",
        "clientId": "dummyPostAuthClientId",
      },
      "region": "dummyPostAuthRegion",
      "request": {
        "newDeviceUsed": false,
        "userAttributes": {
          "cognito:user_status": "EXTERNAL_PROVIDER",
          "email": "dummyPostAuthEmail",
          "email_verified": "false",
          "family_name": "dummyPostAuthFamilyName",
          "given_name": "dummyPostAuthGivenName",
          "identities": "[{\\"userId\\":\\"182347123047012\\",\\"providerName\\":\\"Google\\",\\"providerType\\":\\"Google\\",\\"issuer\\":null,\\"primary\\":true,\\"dateCreated\\":1658487089823}]",
          "picture": "dummyPostAuthPictureUrl",
          "sub": "dummyPostAuthSub",
        },
      },
      "response": {},
      "triggerSource": "PostAuthentication_Authentication",
      "userName": "dummyPostAuthUserName",
      "userPoolId": "dummyPostAuthUserPoolId",
      "version": "dummyPostAuthVersion",
    },
  ],
]
`;

exports[`updateUserFromCognitoEvent > sends "GetCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "Key": {
    "id": "dummyId",
  },
  "TableName": "TP-users",
}
`;

exports[`updateUserFromCognitoEvent > when output from "GetCommand" does not include "Item" > rejects with a correct value 1`] = `[Error: could not find user with id "dummyId"]`;

exports[`updateUserFromCognitoEvent > when user props in event are changed > calls "createDynamoUpdateExpression" with correct args 1`] = `
[
  [
    {
      "dummyChangedProp": "dummyChangedPropValue",
    },
  ],
]
`;

exports[`updateUserFromCognitoEvent > when user props in event are changed > resolves with a correct value 1`] = `
{
  "Attributes": {
    "dummyUpdateCommandAttrKey": "dummyUpdateCommandAttrValue",
  },
}
`;

exports[`updateUserFromCognitoEvent > when user props in event are changed > sends "UpdateCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "ExpressionAttributeNames": "dummyExpressionAttributeNames",
  "ExpressionAttributeValues": "dummyExpressionAttributeValues",
  "Key": {
    "id": "dummyId",
  },
  "TableName": "TP-users",
  "UpdateExpression": "dummyUpdateExpression",
}
`;

exports[`updateUserFromCognitoEvent > when user props in event are not changed > resolves with a correct value 1`] = `undefined`;
